src/main/java/qupath/lib/images/servers/omero/OmeroRawImageServer.java

```diff
@@ -723,6 +730,192 @@ public boolean equals(Object obj) {
+		// get the ROI
+		Roi omeROI = (Roi) roiData.asIObject();
+
+		// get the shapes contained in the ROI (i.e. holes or something else)
+		List<Shape> shapes = omeROI.copyShapes();
+		List<ROI> list = new ArrayList<>();
+
+		// Iterate on shapes, select the correct instance and create the corresponding QuPath ROI
+		if (shapes != null) {
+			Iterator<Shape> it = shapes.iterator();
+
+			while (it.hasNext()){
+				Shape shape = (Shape)it.next();
+
+				if(shape instanceof Rectangle){
+					RectangleData s = new RectangleData(shape);
+					list.add(ROIs.createRectangleROI(s.getX(),s.getY(),s.getWidth(),s.getHeight(),ImagePlane.getPlaneWithChannel(s.getC(), Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Ellipse){
+					EllipseData s = new EllipseData(shape);
+					list.add(ROIs.createEllipseROI(s.getX(),s.getY(),s.getRadiusX(), s.getRadiusY(),ImagePlane.getPlaneWithChannel(s.getC(),Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Point){
+					PointData s = new PointData(shape);
+					list.add(ROIs.createPointsROI(s.getX(),s.getY(),ImagePlane.getPlaneWithChannel(s.getC(),Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Polyline){
+					PolylineData s = new PolylineData(shape);
+					list.add(ROIs.createPolylineROI(s.getPoints().stream().mapToDouble(Point2D.Double::getX).toArray(),
+													s.getPoints().stream().mapToDouble(Point2D.Double::getY).toArray(),
+													ImagePlane.getPlaneWithChannel(s.getC(),Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Polygon){
+					PolygonData s = new PolygonData(shape);
+					list.add(ROIs.createPolygonROI(s.getPoints().stream().mapToDouble(Point2D.Double::getX).toArray(),
+							 					   s.getPoints().stream().mapToDouble(Point2D.Double::getY).toArray(),
+												   ImagePlane.getPlaneWithChannel(s.getC(),Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Label){
+					logger.warn("No ROIs created (requested label shape is unsupported)");
+					//s=new TextData(shape);
+
+				}else if(shape instanceof Line){
+					LineData s = new LineData(shape);
+					list.add(ROIs.createLineROI(s.getX1(),s.getY1(),s.getX2(),s.getY2(),ImagePlane.getPlaneWithChannel(s.getC(),Math.max(s.getZ(), 0), Math.max(s.getT(), 0))));
+
+				}else if(shape instanceof Mask){
+					logger.warn("No ROIs created (requested Mask shape is not supported yet)");
+					//s=new MaskData(shape);
+				}else{
+					logger.warn("Unsupported shape ");
+				}
+			}
+		}
+
+		return list;
+	}
+
+
+	/**
+	 * Output the ROI result of the XOR operation between the 2 input ROIs
+	 *
+	 * @param roi1
+	 * @param roi2
+	 * @return ROI resulting of the XOR operation
+	 */
+	public ROI linkShapes(ROI roi1, ROI roi2){
+		// get the area of the first roi
+		java.awt.geom.AffineTransform at = new java.awt.geom.AffineTransform();
+		at.translate(0, 0);
```

Comment:
A enlever ?

Reply:
at probablememt inutile

Reply:
Yes, it works without any affine transform.